### ISSUE TITLE
Saving and restoring model state

### ISSUE THREAD

**Issue by dschwen** [Argonne-National-Laboratory/neml#95]
For continuation and recovery it would be nice if the internal state of NEML models could be serialized and unserialized.

**Comment by reverendbedford** [comment]
This has been on my list for awhile now.  There are two "modes" that would make sense:

1. The objects dump some serialization of the model parameters.  This wouldn't include the "state" (internal variables, stress, etc.) from the mechanics/materials point of view.  It would let the NEML models be integrated into something like an FE checkpoint file (because the finite element solver is responsible storing the data required to reconstructed the stress, material state, etc).  This would basically be a method to reduce a NEML object back to the XML input, the reverse of the load from XML system.
2. The first option + actually store all the "real" state information.

(1) is much easier than (2).  The problem with (2) is that I've always structured NEML so that the models don't "own" their internal state so that we don't duplicate storage when called from a FE program.

**Comment by dschwen** [comment]
2) is what I'd need. MOOSE has the necessary I/O. So from my perspective it would be _best_ if NEML had an API that exposed the internal state in some clever way. One option would be to serialize/unserialize the state into a raw memory block or a set of vectors (one vector for each type). This ideally would serialize all sub-objects and unserialize their data again. We would not need to construct the models based on this serialization, we'd just fill in state of the freshly constructed objects to warp them back to their state they were in when we serialized it.

**Comment by dschwen** [comment]
Lack of this capability is a bit of a blocker right now. For our mutual friends (no, not the Libyans) we have to simulate over a dozen "phases", which ideally should be separate and distinct MOOSE inputs that each restart off of the previous "phase". Having to do the entire thing in one input and in one go would be super painful.

**Comment by dschwen** [comment]
Or use `std::deque`s instead of vectors and have the signatures be

```
virtual void serialize(std::deque & doubles, std::deque & ints)
{
  _sub_object->serialize(doubles, ints);

  for (std::size_t i = 0; i < _n_hist; ++i)
    doubles.push_back(_history[i]);
}

virtual void unserialize(std::deque & doubles, std::deque & ints)
{
  _sub_object->unserialize(doubles, ints);

  for (std::size_t i = 0; i < _n_hist; ++i)
  {
    _history[i] = doubles.front();
    doubles.pop_front();
  }
}
```

This could be made a bit nicer with two wrapper classes around the deques to have a write only `SerializeBuffer` and a read only `UnserializeBuffer` with pushDouble pushInt popDouble popInt methods (and possibly helpers to push and pop entire vectors -  the vector size could be stored in the int deque)

**Comment by dschwen** [comment]
Hm, I see you have this `History` object, and I see `init_history(History & history)` methods. But sadly at the NEMLModel level this doesn't exist, instead there is `init_hist(double *)`.

That's a bit confusing.

**Comment by dschwen** [comment]
Do you want to have a call about this (requirementsto get functionality we need in MOOSE)? I'd be happy to help implement this.

**Comment by reverendbedford** [comment]
Sure, let's have a call.  I'm busy with another MOOSE-related task earlier this week but I should be able to take a look later in the week or early next week.  This shouldn't be too difficult.

**Comment by reverendbedford** [comment]
Addressed by #118
